src/main/java/spoon/reflect/factory/ModuleFactory.java

```diff
@@ -0,0 +1,163 @@
+import spoon.reflect.declaration.CtModuleRequirement;
+import spoon.reflect.reference.CtPackageReference;
+import spoon.reflect.reference.CtTypeReference;
+import spoon.reflect.visitor.CtVisitor;
+import spoon.support.reflect.declaration.CtElementImpl;
+import spoon.support.reflect.declaration.CtModuleImpl;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ModuleFactory extends SubFactory implements Serializable {
+
+	public static class CtUnnamedModule extends CtModuleImpl {
+		final Map<String, CtModule> allModules = new HashMap<>();
+
+		{
+			this.setSimpleName(CtModule.TOP_LEVEL_MODULE_NAME);
+			this.setParent(new CtElementImpl() {
+				@Override
+				public void accept(CtVisitor visitor) {
+
+				}
+
+				@Override
+				public CtElement getParent() throws ParentNotInitializedException {
+					return null;
+				}
+
+				@Override
+				public Factory getFactory() {
+					return CtUnnamedModule.this.getFactory();
+				}
+			});
+
+			this.setRootPackage(new CtModelImpl.CtRootPackage());
+			this.addModule(this);
+		}
+
+		public boolean addModule(CtModule module) {
+			if (allModules.containsKey(module.getSimpleName())) {
+				return false;
+			}
+			allModules.put(module.getSimpleName(), module);
+			return true;
+		}
+
+		public CtModule getModule(String name) {
+			return allModules.get(name);
+		}
+
+		public Collection<CtModule> getAllModules() {
+			return Collections.unmodifiableCollection(allModules.values());
+		}
+
+		@Override
+		public <T extends CtNamedElement> T setSimpleName(String name) {
+			if (name == null) {
+				return (T) this;
+			}
+
+			if (name.equals(CtModule.TOP_LEVEL_MODULE_NAME)) {
+				return super.setSimpleName(name);
+			}
+
+			return (T) this;
+		}
+
+		@Override
+		public String toString() {
+			return CtModule.TOP_LEVEL_MODULE_NAME;
+		}
+
+		@Override
+		public void accept(CtVisitor visitor) {
+			visitor.visitCtModule(this);
+			getFactory().Module().getAllModules().forEach(module -> {
+				if (!module.getSimpleName().equals(CtModule.TOP_LEVEL_MODULE_NAME)) {
+					module.accept(visitor);
+				}
+			});
+		}
+	}
+
+	public ModuleFactory(Factory factory) {
+		super(factory);
+	}
+
+	public CtUnnamedModule getUnnamedModule() {
+		return (CtUnnamedModule) factory.getModel().getUnnamedModule();
+	}
+
+	public Collection<CtModule> getAllModules() {
+		return getUnnamedModule().getAllModules();
+	}
+
+	public CtModule getModule(String moduleName) {
+		return getUnnamedModule().getModule(moduleName);
+	}
+
+	public CtModule getOrCreate(String moduleName) {
+		if (moduleName == null || moduleName.isEmpty()) {
+			return getUnnamedModule();
+		}
+
+		CtModule ctModule = getUnnamedModule().getModule(moduleName);
+		if (ctModule == null) {
+			ctModule = factory.Core().createModule().setSimpleName(moduleName).setParent(getUnnamedModule());
+			getUnnamedModule().addModule(ctModule);
```

Comment:
What about move of this code `getUnnamedModule().addModule(ctModule);` into `factory.Core().createModule()`. WDYT?